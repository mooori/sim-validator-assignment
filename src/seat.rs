use serde::Serialize;

use crate::validator::Validator;

/// Represents a seat filled by a particular validator. A seat may not outlive the validator it is
/// referrencing.
///
/// The stake corresponding to a seat is set by global configuration and is equal for every seat.
/// Hence, to avoid bloating this struct, it has no field or method to retrieve a seat's stake.
#[derive(Serialize, PartialEq, Debug, Clone)]
pub struct Seat<'validator> {
    validator: &'validator Validator,
}

impl<'validator> Seat<'validator> {
    pub fn new(validator: &'validator Validator) -> Self {
        Self { validator }
    }

    pub fn get_is_malicious(&self) -> bool {
        self.validator.get_is_malicious()
    }
}

pub struct ShuffledSeats<'seats> {
    seats: &'seats mut [Seat<'seats>],
}

// TODO impl new(seats) which shuffles them
impl<'seats> ShuffledSeats<'seats> {
    /// Shuffles the input `seats`.
    pub fn new(seats: &'seats mut [Seat<'seats>]) -> Self {
        // TODO mention library used for randomness in README
        fastrand::shuffle(seats);
        Self { seats }
    }

    pub fn get_seats(&self) -> &[Seat] {
        &self.seats
    }
}
