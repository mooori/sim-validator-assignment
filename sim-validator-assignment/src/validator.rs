use num_rational::Ratio;
use serde::{Deserialize, Serialize};

use crate::config::Config;
use crate::partial_seat::PartialSeat;
use crate::seat::Seat;

#[derive(Deserialize, Serialize, Debug)]
pub struct RawValidatorData {
    pub account_id: String,
    pub stake: u128,
    pub is_malicious: bool,
}

pub fn parse_raw_validator_data(
    config: &Config,
    input: &[RawValidatorData],
) -> (PopulationStats, Vec<Validator>) {
    let mut population_stats = PopulationStats::default();
    let mut validators = vec![];

    for v in input.iter() {
        population_stats.stake += v.stake;
        let num_seats = config.seats_per_stake(v.stake);
        population_stats.seats += num_seats;
        if v.is_malicious {
            population_stats.malicious_stake += v.stake;
            population_stats.malicious_seats += num_seats;
        }
    }

    for v in input.iter() {
        validators.push(Validator {
            account_id: v.account_id.clone(),
            stake: v.stake,
            is_malicious: v.is_malicious,
            num_seats: config.seats_per_stake(v.stake),
            total_stake_share: Ratio::new(v.stake, population_stats.stake),
        })
    }

    (population_stats, validators)
}

impl From<dl_validator_data::ValidatorData> for RawValidatorData {
    /// The returned validator is malicious if `data.is_malicious == Some(true)`, otherwise it is
    /// not malicious.
    fn from(data: dl_validator_data::ValidatorData) -> Self {
        Self {
            account_id: data.account_id,
            stake: data.stake,
            is_malicious: data.is_malicious.is_some_and(|is_malicious| is_malicious),
        }
    }
}

/// Holds data describing a set of validators.
#[derive(Serialize, Default, Debug)]
pub struct PopulationStats {
    /// Sum of validator stakes.
    pub stake: u128,
    /// Sum of stake of malicious validator stakes.
    pub malicious_stake: u128,
    /// Total number of seats of all validators. Note that some stake might not participate in
    /// validation if it is not assigned to a seat. For example, if a validator has a stake of 17
    /// and seat price is 5, then the validator holds 3 seats (worth 15 stake) and 2 stake remain
    /// unassigned.
    pub seats: u64,
    /// The number of seats held by malicious validators.
    pub malicious_seats: u64,
}

#[derive(Serialize, PartialEq, Debug)]
pub struct Validator {
    account_id: String,
    stake: u128,
    is_malicious: bool,
    num_seats: u64,
    total_stake_share: Ratio<u128>,
}

impl Validator {
    /// Returns the validator's seats. The number of seats a validator claims is determined by the
    /// stake required per seat and the validator's stake.
    pub fn seats(&self) -> Vec<Seat> {
        let seat = Seat::new(self);
        vec![seat.clone(); self.num_seats_as_usize()]
    }

    /// Returns `num_seats` as usize and panics on overflow.
    ///
    /// Field `num_seats` is an unsigned integer in the struct to facilitate calculations.
    pub fn num_seats_as_usize(&self) -> usize {
        usize::try_from(self.num_seats).expect("num_seats should fit into usize")
    }

    /// Returns the validator's `PartialSeat`. It might have a weight of 0 if the validator's
    /// stake is 0 or there is no remainder after distributing the validator's stake to full seats.
    ///
    /// # Panics
    ///
    /// Panics if `stake_per_seat` is 0.
    pub fn partial_seat(&self, stake_per_seat: u128) -> PartialSeat {
        let weight = self.stake % stake_per_seat;
        PartialSeat::new(self, weight)
    }

    pub fn get_is_malicious(&self) -> bool {
        self.is_malicious
    }
}

pub fn new_ordered_seats(validators: &[Validator]) -> Vec<Seat> {
    // The `i`th element holds the seats of `validators[i]`
    let seats_per_validator: Vec<Vec<Seat>> = validators.iter().map(|v| v.seats()).collect();

    seats_per_validator.into_iter().flatten().collect()
}

/// Constructs `PartialSeats` for the provided `validators`. A validator whose partial seat has a
/// weight of zero will not be represented in the returned vector.
pub fn new_ordered_partial_seats(
    validators: &[Validator],
    stake_per_seat: u128,
) -> Vec<PartialSeat> {
    validators
        .iter()
        .filter_map(|v| {
            let partial_seat = v.partial_seat(stake_per_seat);
            if partial_seat.get_weight() > 0 {
                Some(partial_seat)
            } else {
                None
            }
        })
        .collect()
}

#[cfg(test)]
pub mod tests {
    use num_rational::Ratio;

    use crate::partial_seat::PartialSeat;
    use crate::validator::new_ordered_partial_seats;

    use super::Config;
    use super::RawValidatorData;
    use super::Validator;
    use super::{new_ordered_seats, parse_raw_validator_data};

    fn new_test_validator() -> Validator {
        Validator {
            account_id: "validator_0".to_owned(),
            stake: 300,
            is_malicious: false,
            num_seats: 3,
            total_stake_share: Ratio::new(3, 100),
        }
    }

    pub fn new_test_raw_validator_data() -> Vec<RawValidatorData> {
        vec![
            // Stake covers a full number of seats.
            RawValidatorData {
                account_id: "validator_0".to_owned(),
                stake: 500,
                is_malicious: false,
            },
            // Stake is elligible for seats and there is a remainder (partial seat).
            RawValidatorData {
                account_id: "validator_1".to_owned(),
                stake: 310,
                is_malicious: true,
            },
            // Stake is not sufficient for a seat.
            RawValidatorData {
                account_id: "validator_2".to_owned(),
                stake: 90,
                is_malicious: false,
            },
            // Some more validators, to have enough for running tests.
            RawValidatorData {
                account_id: "validator_3".to_owned(),
                stake: 100,
                is_malicious: true,
            },
            RawValidatorData {
                account_id: "validator_4".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_5".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_6".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_7".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_8".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_9".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_10".to_owned(),
                stake: 100,
                is_malicious: false,
            },
            RawValidatorData {
                account_id: "validator_11".to_owned(),
                stake: 100,
                is_malicious: false,
            },
        ]
    }

    #[test]
    fn test_new_ordered_seats() {
        assert_eq!(new_ordered_seats(&[]), vec![]);

        let config = Config::new_mock(false);
        let (_, validators) = parse_raw_validator_data(&config, &new_test_raw_validator_data());
        // Use a small set of validators to avoid bloating snapshot files.
        let validators = &validators[0..3];
        insta::with_settings!({
            info => &(
                config,
                "validators:",
                validators
            ),
        }, {
            insta::assert_yaml_snapshot!(new_ordered_seats(validators));
        })
    }

    #[test]
    fn test_parse_raw_validator_input() {
        let config = Config::new_mock(false);
        let (population_stats, validators) =
            parse_raw_validator_data(&config, &new_test_raw_validator_data());

        insta::with_settings!({
            info => &config,
        }, {
            insta::assert_yaml_snapshot!(population_stats);
            insta::assert_yaml_snapshot!(validators);
        })
    }

    #[test]
    fn test_validator_seats() {
        let mut validator_0_seats = new_test_validator();
        validator_0_seats.num_seats = 0;
        insta::assert_yaml_snapshot!(validator_0_seats.seats());

        let validator_3_seats = new_test_validator();
        insta::assert_yaml_snapshot!(validator_3_seats.seats());
    }

    #[test]
    fn test_validator_partial_seat() {
        let validator = new_test_validator();

        // stake fills full seats without remainder
        assert_eq!(validator.partial_seat(100), PartialSeat::new(&validator, 0));

        // there is a partial seat with `weight > 0`
        assert_eq!(validator.partial_seat(90), PartialSeat::new(&validator, 30));
    }

    #[test]
    fn test_new_ordered_partial_seats() {
        assert_eq!(new_ordered_partial_seats(&[], 1), vec![]);

        let config = Config::new_mock(true);
        let (_, validators) = parse_raw_validator_data(&config, &new_test_raw_validator_data());
        insta::with_settings!({
            info => &(
                &config,
                "validators:",
                &validators
            ),
        }, {
            insta::assert_yaml_snapshot!(new_ordered_partial_seats(&validators, config.stake_per_seat));
        })
    }
}
