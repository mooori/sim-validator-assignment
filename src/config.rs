use clap::Parser;
use num_rational::Ratio;
use serde::Serialize;

use crate::seat::Seat;

// TODO change stake related types to u128
#[derive(Parser, Serialize, Debug)]
pub struct Config {
    #[arg(long)]
    pub num_blocks: u64,
    // TODO change to `u16`
    #[arg(long)]
    pub num_shards: u64,
    /// TODO mention excess validators are ignored/unassigned.
    #[arg(long)]
    pub seats_per_shard: u64,
    /// The amount of stake required to get one seat.
    #[arg(long)]
    pub stake_per_seat: u64,
    /// If the ratio of malicious stake is higher than this threshold, the shard is considered
    /// corrupted, i.e. a security failure occured.
    #[arg(long)]
    pub max_malicious_stake_per_shard: Ratio<u64>,
}

impl Config {
    #[cfg(test)]
    pub fn new_mock() -> Self {
        Self {
            num_blocks: 1_000,
            num_shards: 4,
            seats_per_shard: 2,
            stake_per_seat: 100,
            max_malicious_stake_per_shard: Ratio::new(1, 3),
        }
    }

    /// Returns the number of (full) seats that can be claimed by `stake`.
    pub fn seats_per_stake(&self, stake: u64) -> u64 {
        // Integer division in Rust returns the floor as described here
        // https://doc.rust-lang.org/std/primitive.u64.html#method.div_euclid
        stake / self.stake_per_seat
    }

    /// Returns the amount of seats for all shards that must be filled by validators.
    pub fn total_seats(&self) -> u64 {
        self.num_shards
            .checked_mul(self.seats_per_shard)
            .expect("min_required_seats should fit into return type")
    }

    /// Collects the (consecutive) seats required for `shard_idx` starting from `seats[shard_idx *
    /// self.seats_per_shard]`.
    ///
    /// # Panics
    ///
    /// If `shard_idx` is out of bounds or if `seats` contains not enough seats.
    pub fn collect_seats_for_shard<'seats>(
        &self,
        shard_idx: usize,
        seats: &'seats [Seat],
    ) -> anyhow::Result<Vec<&Seat<'seats>>> {
        if u64::try_from(shard_idx).unwrap() >= self.num_shards {
            anyhow::bail!(
                "shard_idx {} is an invalid index for {} shards",
                shard_idx,
                self.num_shards
            )
        }
        let required_seats = usize::try_from(self.num_shards * self.seats_per_shard).unwrap();
        if seats.len() < required_seats {
            anyhow::bail!(
                "validators fill only {}/{} of seats",
                seats.len(),
                required_seats
            );
        }

        let seats_per_shard = usize::try_from(self.seats_per_shard).unwrap();
        let start = shard_idx * seats_per_shard;
        let shard_seats: Vec<_> = seats[start..start + seats_per_shard].iter().collect();

        Ok(shard_seats)
    }
}

#[cfg(test)]
mod tests {
    use super::Config;
    use crate::validator::tests::new_test_raw_validator_data;
    use crate::validator::{new_ordered_seats, parse_raw_validator_data};

    #[test]
    pub fn test_seats_per_stake() {
        let config = Config::new_mock();
        assert_eq!(config.seats_per_stake(20), 0);
        assert_eq!(config.seats_per_stake(100), 1);
        assert_eq!(config.seats_per_stake(530), 5);
    }

    #[test]
    pub fn test_total_seats() {
        let config = Config::new_mock();
        assert_eq!(config.total_seats(), 8);
    }

    #[test]
    pub fn test_collect_seats_for_shard() {
        let config = Config::new_mock();
        let (_, validators) = parse_raw_validator_data(&config, &new_test_raw_validator_data());
        // Using ordered seats as input to have a deterministic result of `collect_seats_for_shard`.
        let seats = new_ordered_seats(&validators);

        insta::with_settings!({
            info => &(
                &config,
                "seats:",
                &seats
            )
        }, {
            insta::assert_yaml_snapshot!(config.collect_seats_for_shard(0, &seats).unwrap());
            insta::assert_yaml_snapshot!(config.collect_seats_for_shard(1, &seats).unwrap());
            insta::assert_yaml_snapshot!(config.collect_seats_for_shard(2, &seats).unwrap());
            insta::assert_yaml_snapshot!(config.collect_seats_for_shard(3, &seats).unwrap());
        })
    }

    #[test]
    pub fn test_collect_seats_for_shard_errors() {
        let config = Config::new_mock();
        let (_, validators) = parse_raw_validator_data(&config, &new_test_raw_validator_data());
        let seats = new_ordered_seats(&validators);

        insta::assert_debug_snapshot!(config.collect_seats_for_shard(4, &seats));
        insta::assert_debug_snapshot!(config.collect_seats_for_shard(0, &[]));
    }
}
