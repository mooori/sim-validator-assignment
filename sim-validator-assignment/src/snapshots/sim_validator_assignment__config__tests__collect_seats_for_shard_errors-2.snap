---
source: src/config.rs
expression: "config.collect_seats_for_shard(0, &[])"
---
Err(
    "validators fill only 0/8 of seats",
)
